Updated ps_get_owner() to return a stringified uid if process owner name is not set.

rbuskens opened the pull request.

igorpeshansky commented on src/processes.c:
```diff
@@ -1529,8 +1529,13 @@ static char *ps_get_owner(pid_t pid)
+        if (passwd_result) {
             result = sstrdup (passwd_result->pw_name);
+        } else {
+	    char temp_str[6];  // 5 digits plus string terminator.
```
These look misaligned. Stray tab characters?

igorpeshansky replied:
Can we get a better name than `temp_str`? E.g., `uid_str`?

rbuskens replied:
Fixed.

rbuskens replied:
Yes, done.

igorpeshansky commented on src/processes.c:
```diff
@@ -1529,8 +1529,13 @@ static char *ps_get_owner(pid_t pid)
             result = sstrdup (passwd_result->pw_name);
+        } else {
+	    char temp_str[6];  // 5 digits plus string terminator.
+	    snprintf(temp_str, sizeof(temp_str), "%d", uid);
```
`"%u"` -- `uid_t` is unsigned on Linux.

rbuskens replied:
Fixed.

igorpeshansky commented on src/processes.c:
```diff
@@ -1529,8 +1529,13 @@ static char *ps_get_owner(pid_t pid)
+        if (passwd_result) {
             result = sstrdup (passwd_result->pw_name);
+        } else {
+            char uid_str[6];  // 5 digits plus string terminator.
```
Let's add a comment here, e.g.:
```
// Send the numeric uid if name is not available.
```

rbuskens replied:
Done.

kosak commented on src/processes.c:
```diff
@@ -1529,8 +1529,14 @@ static char *ps_get_owner(pid_t pid)
             result = sstrdup (passwd_result->pw_name);
+        } else {
+	    // Send the numeric uid if name is not available.
+            char uid_str[6];  // 5 digits plus string terminator.
```
This buffer is too small. I'm not aware of any guarantee that user ids will fit in 16 bits, which is what this code seems to be assuming.

rbuskens replied:
Looked around online and information about uid_t does not seem to be current. So, assumed a worst case of 64 bits and adjusted the code accordingly.

igorpeshansky replied:
(a) The buffer was perfectly fine — because `snprintf` was used, the value might be truncated, but would never overflow.
(b) `strtoul` produces a 32-bit value, and the `"%lu"` format assumes that as well, so `%lu` will never result in more than 10 decimal digits.
(c) If you were worried that `uid_t` would be larger than 32 bits in some cases, you could have used something like `sizeof(uid_t)/3+1` as the static size of the array.

kosak commented on src/processes.c:
```diff
@@ -1529,8 +1529,14 @@ static char *ps_get_owner(pid_t pid)
+        } else {
+	    // Send the numeric uid if name is not available.
+            char uid_str[6];  // 5 digits plus string terminator.
+            snprintf(uid_str, sizeof(uid_str), "%u", uid);
```
As a technical matter, ```%u``` is not the right format specifier for a variable of type ```int```.

I think, due to the issue I raised above (the code not knowing the size of user ids), I think I'd recommend using a bigger type (namely ```unsigned long```) in line 1520, then using ```strtoul``` rather than atoi in line 1529, and then using ```%ul``` here.

I will also note how funny it looks to take a string, parse it into an integer, and then convert that integer back to a string. But that's probably tolerable because it's the shortest version of this code.

rbuskens replied:
Used uid_t as the type and took your suggestion of using strtoul.

Re: string --> integer --> string: extracting the uid from /proc/<pid>/status returns a string with multiple numbers in it. The atoi/strtoul grabs the first one. Otherwise would have avoided what looks to be a redundant conversion.

igorpeshansky replied:
With `atoi`, the double conversion is unavoidable. However, `strtoul` allows one to save the end pointer in the string, so you could use something like the following instead:
```
char* uid_end;
uid = strtoul (line + 5, &uid_end, /* base */ 10);
...
} else {
  result = strndup (line + 5, uid_end - line - 5);
}
```
Or we could even add an `sstrndup` function upstream for portability and use that.

igorpeshansky replied:
Better yet, since we don't care about retaining the original line:
```
char* uid_end;
uid = strtoul (line + 5, &uid_end, /* base */ 10);
...
} else {
  *uid_end = '\0';
  result = sstrdup (line + 5);
}
```

kosak commented on src/processes.c:
```diff
@@ -1526,15 +1526,15 @@ static char *ps_get_owner(pid_t pid)
 	    // Send the numeric uid if name is not available.
             char uid_str[21];  // Worst case is 20 digits plus string terminator.
-            snprintf(uid_str, sizeof(uid_str), "%llu", uid);
+            snprintf(uid_str, sizeof(uid_str), "%lu", uid);
```
Not quite unfortunately. The issue here is that the format specifier needs to match the type of the argument. ```%lu``` needs an ```unsigned long```, but you are passing a ```uid_t``` (which may or may not be an unsigned long).  I think the shortest way to deal with this is

```
snprintf(uid_str, sizeof(uid_str), "%lu", (unsigned long)uid);
```

kosak commented:
In this interface I can't find igor's comment to reply to, so I'll comment right here.
* *The buffer was perfectly fine — because snprintf was used, the value might be truncated, but would never overflow.*
Indeed. I was worried about truncation. I never said I was worried about overflow. My change was intended to fix the truncation problem. Why do you not care about the truncation problem?
* *strtoul produces a 32-bit value, and the "%lu" format assumes that as well, so %lu will never result in more than 10 decimal digits.*
No, the format specifiers do not talk about bits. ```%lu``` means ```unsigned long``` and it needs to take an argument of type unsigned long. It's not correct to muddle C types with the bit lengths that they happen to be defined as right now in 2017 on your platform. It is also not correct, to assume that things will work because your format specifier represents a type "no smaller than" the size of your argument. It's not even cool, from a pedantic perspective, so far as I know, to pass a signed type to a format specifier that expects an unsigned type of the same size. And by the way, even by that argument, since ```%lu``` is ```unsigned long```, and since ```unsigned long``` is going to be 64 bits on the machine you're sitting at, I don't know why you're talking about 32 bits.
* *If you were worried that uid_t would be larger than 32 bits in some cases, you could have used something like sizeof(uid_t)/3+1 as the static size of the array.*
Yes, I was trying to defend against that, but also was trying to not make the code look excessively pedantic. So the point of my change, was to use a "worst case" type like unsigned long which would give lots of breathing room (we will be dead before sizeof(uid_t) > sizeof(unsigned long)). Put another way, I felt the original size (5 + 1) was definitely not future-proof, but assuming unsigned long would be good enough. A solution like ```char buffer[sizeof(uid_t)/3+1]``` is also ok with me, though I admit it seems a little pedantic (there's no performance advantage to getting the buffer size exactly right). But I would be fine with that and also fine with ```char buffer[128]```.

I think it's a little chaotic to whipsaw this code back and forth like this. I think the ```strtoul``` with unsigned long was the best version. I don't know why you (igor) would recommend going back to the ```atoi``` version. I'm happy to continue to discuss this if my input is welcome (I'm also happy with the opposite of that statement) so let me know if I should continue. But personally I'd rather discuss it as a discussion, rather than the ObamaCare style repeal-and-replace approach that we're using now. 😸

kosak commented:
PS I like @igorpeshansky's suggestion
```result = strndup (line + 5, uid_end - line - 5);```
I think that's the awesomest because it avoids all this snprintf and buffer size drama altogether.

rbuskens commented:
Reverted back to using unsigned long as a worst-case assumption to give us lots of breathing room / future-proofing. Updated the code to avoid the buffer size drama altogether.

igorpeshansky commented:
@kosak For the record, I've never suggested reverting to `atoi`. :-) I was also typing these on a phone while riding on a crowded train, so some context got mixed up (e.g., the "you" in "if you were worried..." referred to Rick and not Corey). I think the whole "repeal-and-replace" exchange was Rick being somewhat overeager to apply every comment as it appeared, rather than waiting for the dust to settle. That said, your input is definitely appreciated.

igorpeshansky commented on src/processes.c:
```diff
@@ -1526,11 +1527,16 @@ static char *ps_get_owner(pid_t pid)
+        if (passwd_result) {
             result = sstrdup (passwd_result->pw_name);
+        } else {
+	    // Send the numeric id if name is not available.
```
More stray tabs (in this line and next).

igorpeshansky replied:
Nit: s/numeric id/numeric uid/.

rbuskens replied:
Fixed.

rbuskens replied:
Fixed.

rbuskens commented:
+1 to what Igor wrote, Corey!

kosak commented:
Ah, ok. Thanks Igor. (and I wish I knew how to reply to a comment here). I definitely was getting confused by all the changes (and my aging brain only holds about 3 bits of information any more), so thanks for the clarification! And I'm always happy to drop in via parachute.

igorpeshansky commented:
LGTM :shipit:
